source/beamline/AMMotorGroup.cpp

```diff
@@ -312,17 +805,59 @@ AMMotorGroup::AMMotorGroup(QObject *parent)
 
 }
 
-AMMotorGroupObject *AMMotorGroup::motorGroupObject(const QString &name) const
+QList<QString> AMMotorGroup::names() const
+{
+	return groupObjects_.keys();
+}
+
+QList<AMMotorGroupObject *> AMMotorGroup::motorGroupObjects() const
+{
+	return groupObjects_.values();
+}
+
+int AMMotorGroup::size() const
+{
+	return groupObjects_.size();
+}
+
+AMMotorGroupObject * AMMotorGroup::motorGroupObject(const QString &name) const
 {
-	return infoMap_.value(name, 0);
+	return groupObjects_.value(name, 0);
 }
 
 QString AMMotorGroup::name(AMMotorGroupObject *motorObject) const
 {
-	return infoMap_.key(motorObject, QString());
+	return groupObjects_.key(motorObject, QString());
 }
 
-void AMMotorGroup::addMotorGroupObject(const QString &name, AMMotorGroupObject *object)
+void AMMotorGroup::addMotorGroupObject(AMMotorGroupObject *object)
 {
-	infoMap_.insert(name, object);
+	if(object == 0) {
+		return;
+	}
+
+	AMMotorGroupObject* currentObject = motorGroupObject(object->name());
+
+	if(currentObject == object) {
```

Comment:
Okay, here's how I'm thinking this will go for resource ownership, in that case:
- AMMotorGroup takes no ownership of it's contained AMMotorGroupObjects. The above addMotorGroupObject(...) function now returns a reference to an AMMotorGroupObect that has been replaced, or 0 if no replacement took place.
- AMMotorGroupObjectss still take ownership of their three AMMotorGroupAxes. This class is so closely tied to the motor group object that it would be an odious overhead to have the user manually keep track of these resources separate from the group object.
- Similarly AMMotorGroupAxes take ownership of their AMMotorGroupMotions.
- None of these classes takes any ownership of the contained AMControl's. After all these are just convenience organizational classes.